dot/state/pruner.go

```diff
@@ -0,0 +1,363 @@
+	jr := newJournalRecord(blockHash, inserted, deleted)
+
+	key := &journalKey{blockNum, blockHash}
+	err := p.storeJournal(key, jr)
+	if err != nil {
+		return fmt.Errorf("failed to store journal record for %d: %w", blockNum, err)
+	}
+
+	logger.Debug("journal record stored", "block num", blockNum)
+	p.addDeathRow(jr, blockNum)
+	return nil
+}
+
+func (p *fullNodePruner) addDeathRow(jr *journalRecord, blockNum int64) {
+	if blockNum == 0 {
+		return
+	}
+
+	p.Lock()
+	defer p.Unlock()
+
+	// The block is already pruned.
+	if blockNum < p.pendingNumber {
+		return
+	}
+
+	p.processInsertedKeys(jr.insertedKeys, jr.blockHash)
+
+	// add deleted keys from journal to death index
+	for _, k := range jr.deletedKeys {
+		p.deathIndex[k] = blockNum
+	}
+
+	deletedKeys := make(map[common.Hash]int64)
+	for _, data := range jr.deletedKeys {
+		deletedKeys[data] = blockNum
+	}
+
+	blockIndex := blockNum - p.pendingNumber
+	for idx := blockIndex - int64(len(p.deathList)); idx >= 0; idx-- {
+		p.deathList = append(p.deathList, deathRow{})
+	}
+
+	record := &deathRecord{
+		blockHash:   jr.blockHash,
+		deletedKeys: deletedKeys,
+	}
+
+	// add deathRow to deathList
+	p.deathList[blockIndex] = append(p.deathList[blockIndex], record)
+}
+
+// Remove re-inserted keys
+func (p *fullNodePruner) processInsertedKeys(insKeys []common.Hash, blockHash common.Hash) {
+	for _, k := range insKeys {
+		num, ok := p.deathIndex[k]
+		if !ok {
+			continue
+		}
+		records := p.deathList[num-p.pendingNumber]
+		for _, v := range records {
+			if v.blockHash == blockHash {
+				delete(v.deletedKeys, k)
+			}
+		}
+		delete(p.deathIndex, k)
+	}
+}
+
+func (p *fullNodePruner) start() {
+	logger.Info("pruning started")
+
+	var canPrune bool
+	checkPruning := func() {
+		p.Lock()
+		defer p.Unlock()
+		if int64(len(p.deathList)) <= p.retainBlocks {
+			canPrune = false
+			return
+		}
+		canPrune = true
+
+		// pop first element from death list
+		row := p.deathList[0]
+		blockNum := p.pendingNumber
+
+		logger.Debug("pruning block", "block num", blockNum)
+
+		for _, record := range row {
+			err := p.deleteKeys(record.deletedKeys)
+			if err != nil {
+				logger.Warn("failed to prune keys", "block num", blockNum, "error", err)
+				return
+			}
+
+			for k := range record.deletedKeys {
+				delete(p.deathIndex, k)
+			}
+		}
+
+		err := p.storeLastPrunedIndex(blockNum)
+		if err != nil {
+			logger.Error("failed to store last pruned index", "block num", blockNum, "error", err)
+			return
+		}
+
+		p.deathList = p.deathList[1:]
+		p.pendingNumber++
+
+		for _, record := range row {
+			jk := &journalKey{blockNum, record.blockHash}
+			err = p.deleteJournalRecord(jk)
+			if err != nil {
+				logger.Error("failed to delete journal record", "block num", blockNum, "error", err)
+				return
+			}
+		}
+	}
+
+	for {
+		checkPruning()
+		// Don't sleep if we have data to prune.
+		if !canPrune {
+			time.Sleep(pruneInterval)
```

Comment:
using time.After will allow this goroutine to go move into a blockign state.
```suggestion
			<- time.After(pruneInterval).C
```

Reply:
I did some benchmarking around this. `time.After` seems to have the worst performance CPU and Memory wise.
https://github.com/ChainSafe/gossamer/issues/1623#issue-911375689
Created an issue and report here.

Reply:
Necro-comment-reader right here, both statements do the exact same really, maybe except time.After spins a goroutine whilst time.Sleep hooks into the OS/kernel directly.